controller/tap/server.go

```diff
@@ -90,7 +117,55 @@ func (s *server) Tap(req *public.TapRequest, stream pb.Tap_TapServer) error {
 }
 
 func (s *server) TapByResource(req *public.TapByResourceRequest, stream pb.Tap_TapByResourceServer) error {
-	return fmt.Errorf("unimplemented")
+	if req == nil {
+		return errors.New("TapByResource received nil TapByResourceRequest")
+	}
+	if req.Target == nil {
+		return fmt.Errorf("TapByResource received nil target ResourceSelection: %+v", *req)
+	}
+
+	pods, err := s.getPodsFor(*req.Target)
+	if err != nil {
+		return err
+	}
+
+	if len(pods) == 0 {
+		return fmt.Errorf("No pods found for ResourceSelection: %+v", *req.Target)
+	}
+
+	log.Infof("Tapping %d pods for target: %+v", len(pods), *req.Target.Resource)
+
+	events := make(chan *common.TapEvent)
+
+	go func() { // Stop sending back events if the request is cancelled
+		<-stream.Context().Done()
+		close(events)
+	}()
+
+	// divide the rps evenly between all pods to tap
+	rpsPerPod := req.MaxRps / float32(len(pods))
+	if rpsPerPod < 1 {
+		rpsPerPod = 1
+	}
+
+	match, err := makeByResourceMatch(req.Match)
+	if err != nil {
+		return nil
```

Comment:
is this correct? What is this condition?

Reply:
Nice catch, this is a bug. I've added a test to catch this in case it regresses. Looks like we've had a similar bug in the original `Tap`, that is fixed as well, but no tests added as we'll be removing soon.